reps/2023-06-30-serve-cpp-deployment.md

```diff
@@ -0,0 +1,411 @@
+#include <algorithm>
+
+namespace ray {
+namespace serve {
+
+class RankService {
+ public:
+  std::vector<std::string> Rank(std::unordered_map<std::string, float> recommends, int num) {
+    std::priority_queue<std::pair<std::string, float>, std::vector<std::pair<std::string, float>>, std::greater<std::pair<std::string, float>>> pq;
+    for (auto& pair : recommends) {
+        pq.push(pair);
+        if (pq.size() > num) {
+            pq.pop();
+        }
+    }
+
+    std::vector<std::string> result;
+    while (!pq.empty()) {
+        result.push_back(pq.top().first);
+        pq.pop();
+    }
+    std::reverse(result.begin(), result.end());
+    return result;
+  }
+
+  static RankService *FactoryCreate() {
+    return new RankService();
+  }
+};
+
+// Register function
+SERVE_DEPLOYMENT(RankService::FactoryCreate);
+
+}  // namespace serve
+}  // namespace ray
+
+```
+RecommendService is a sequential invocation of other services without complex processing logic, so we can directly use the DAG ability to connect these services, eliminating the need for RecommendService and simplifying user logic. 
+Next, we start the Ray Serve runtime and use Python Serve API deploy these Service as Deployment: 
+```python
+feature_service = serve.deployment(_func_or_class='FeatureService::FactoryCreate', name='feature_service', language='CPP')
+similarity_service = serve.deployment(_func_or_class='SimilarityService::FactoryCreate', name='similarity_service', language='CPP')
+rank_service = serve.deployment(_func_or_class='RankService::FactoryCreate', name='rank_service', language='CPP')
+
+with InputNode() as input:
+    features = feature_service.GetVector.bind(input[0])
+    similarities = similarity_service.GetSimilarity.bind(features)
+    rank_result = rank_service.Rank.bind(similarities, input[1])
+
+graph = DAGDriver.bind(rank_result, http_adapter=json_request)
+handle = serve.run(graph)
+ref = handle.GetVector.remote()
+result = ray.get(ref)
+print(result)
+```
+### Calling Ray Serve Deployment with HTTP
+```java
+curl http://127.0.0.1:8000?request=computer&num=1
+```
+### Overall Design
+   ![image](https://user-images.githubusercontent.com/40862361/250100276-5eeed3a1-70a2-4208-8563-e0b1bf7f7fd2.png)  
+Ray Serve maintains Controller Actor and Ingress Actor. So these two roles are not related to the language of the user's choice. And they have the ability to manage cross-language deployments and route requests.
+### C++ Case Deduction
+The businesses can send control commands to the Controller Actor of Ray Serve, which includes creating ingress, creating deployment, etc. When publishing C++ online services, the DeploymentState component needs to create C++ Deployment  Actors. Users can call their business logic in Python/Java Driver, Ray task, or Ray Actor, and the requests will be dispatched to C++ Deployment Actors.
+![image](https://user-images.githubusercontent.com/40862361/250100685-96466519-1f8e-414e-8d8a-1bcb63952a41.png)
```

Comment:
Will this use the existing Ray C++ API? Do there already exist cross-language APIs for interacting with C++ actors?

Reply:
Yes, using the cross-language capabilities of C++ Worker